pmemcheck/pmc_main.c

```diff
@@ -296,18 +302,193 @@ is_pmem_access(Addr addr, SizeT size)
 }
 
 /**
+* \brief State to string change for information purposes.
+*/
+static const char *
+store_state_to_string(enum store_state state)
+{
+    switch (state) {
+        case STST_CLEAN:
+            return "CLEAN";
+        case STST_DIRTY:
+            return "DIRTY";
+        case STST_FLUSHED:
+            return "FLUSHED";
+        case STST_FENCED:
+            return "FENCED";
+        case STST_COMMITTED:
+            return "COMMITTED";
+        default:
+            return NULL;
+    }
+}
+
+/**
+ * \brief Prints registered flush errors.
+ */
+static void
+print_flush_errors(void)
+{
+    if (pmem.flush_errors_reg) {
+        VG_(umsg)("\nNumber of multiply flushed stores: %lu\n",
+                pmem.flush_errors_reg);
+        VG_(umsg)("Stores flushed multiple times:\n");
+        struct pmem_st *tmp;
+        Int i;
+        for (i = 0; i < pmem.flush_errors_reg; ++i) {
+            tmp = pmem.flush_errors[i];
+            VG_(umsg)("[%d] ", i);
+            VG_(pp_ExeContext)(tmp->context);
+            VG_(printf)("\tAddress: 0x%lx\tsize: %llu\tstate: %s\n",
+                    tmp->addr, tmp->size, store_state_to_string(tmp->state));
+        }
+    }
+}
+
+/**
+ * \brief Prints registered flush misses.
+ */
+static void
+print_flush_misses(void)
+{
+    if (pmem.flush_misses_reg) {
+        VG_(umsg)("\nNumber of unnecessary flushes: %lu\n",
+                pmem.flush_misses_reg);
+        struct pmem_st *tmp;
+        Int i;
+        for (i = 0; i < pmem.flush_misses_reg; ++i) {
+            tmp = pmem.flush_misses[i];
+            VG_(umsg)("[%d] ", i);
+            VG_(pp_ExeContext)(tmp->context);
+            VG_(printf)("\tAddress: 0x%lx\tsize: %llu\n", tmp->addr, tmp->size);
+        }
+    }
+}
+
+/**
+ * \brief Prints registered multiple stores.
+ */
+static void
+print_multiple_stores(void)
+{
+    if (pmem.track_multiple_stores && (pmem.multiple_stores_reg > 0)) {
+        VG_(umsg)("\nNumber of overwritten stores: %lu\n",
+                pmem.multiple_stores_reg);
+        VG_(umsg)("Overwritten stores before they were made persistent:\n");
+        struct pmem_st *tmp;
+        Int i;
+        for (i = 0; i < pmem.multiple_stores_reg; ++i) {
+            tmp = pmem.multiple_stores[i];
+            VG_(umsg)("[%d] ", i);
+            VG_(pp_ExeContext)(tmp->context);
+            VG_(printf)("\tAddress: 0x%lx\tsize: %llu\tstate: %s\n",
+                    tmp->addr, tmp->size, store_state_to_string(tmp->state));
+        }
+    }
+}
+
+/**
+ * \brief Prints registered store statistics.
```

Comment:
Are these not flushed or not drained (not committed), or both?  Can we distinguish them?  Does it make sense?